src/signTxOutput.c

```diff
@@ -0,0 +1,664 @@
+
+		advanceState();
+	}
+	UI_STEP_END(HANDLE_TOKEN_GROUP_STEP_INVALID);
+}
+
+static void signTxOutput_handleTokenGroupAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		// sanity checks
+		CHECK_STATE(STATE_OUTPUT_TOKEN_GROUP);
+
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		token_group_t* tokenGroup = &subctx->stateData.tokenGroup;
+
+		// parse data
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		VALIDATE(view_remainingSize(&view) >= MINTING_POLICY_ID_SIZE, ERR_INVALID_DATA);
+		STATIC_ASSERT(SIZEOF(tokenGroup->policyId) == MINTING_POLICY_ID_SIZE, "inconsistent policy id size");
+		view_memmove(tokenGroup->policyId, &view, MINTING_POLICY_ID_SIZE);
+
+		VALIDATE(view_remainingSize(&view) == 4, ERR_INVALID_DATA);
+		uint32_t numTokenAmounts = parse_u4be(&view);
+		VALIDATE(numTokenAmounts <= OUTPUT_TOKEN_GROUP_NUM_TOKENS_MAX, ERR_INVALID_DATA);
+		STATIC_ASSERT(OUTPUT_TOKEN_GROUP_NUM_TOKENS_MAX <= UINT16_MAX, "wrong max token amounts in a group");
+		ASSERT_TYPE(subctx->numTokenAmounts, uint16_t);
+		subctx->numTokenAmounts = (uint16_t) numTokenAmounts;
+	}
+
+	security_policy_t policy = policyForSignTxOutputTokenGroup(&subctx->stateData.tokenGroup);
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+
+	{
+		// select UI step
+		switch (policy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {subctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_TOKEN_GROUP_STEP_DISPLAY);
+			CASE(POLICY_ALLOW_WITHOUT_PROMPT, HANDLE_TOKEN_GROUP_STEP_RESPOND);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+	}
+
+	{
+		// add tokengroup to tx
+		TRACE("Adding token group hash to tx hash");
+		txHashBuilder_addOutput_tokenGroup(
+		        txHashBuilder,
+		        subctx->stateData.tokenGroup.policyId, MINTING_POLICY_ID_SIZE,
+		        subctx->numTokenAmounts
+		);
+		TRACE();
+	}
+
+	signTxOutput_handleTokenGroup_ui_runStep();
+}
+
+// ============================== TOKEN AMOUNT ==============================
+
+enum {
+	HANDLE_TOKEN_AMOUNT_STEP_DISPLAY_NAME = 800, // TODO
+	HANDLE_TOKEN_AMOUNT_STEP_DISPLAY_AMOUNT,
+	HANDLE_TOKEN_AMOUNT_STEP_RESPOND,
+	HANDLE_TOKEN_AMOUNT_STEP_INVALID,
+};
+
+static void signTxOutput_handleTokenAmount_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	ui_callback_fn_t* this_fn = signTxOutput_handleTokenAmount_ui_runStep;
+
+	UI_STEP_BEGIN(subctx->ui_step);
+
+	UI_STEP(HANDLE_TOKEN_AMOUNT_STEP_DISPLAY_NAME) {
+		token_amount_t* tokenAmount = &subctx->stateData.tokenAmount;
+		if (str_isTextPrintable(tokenAmount->assetName, tokenAmount->assetNameSize)) {
+			char name[ASSET_NAME_SIZE_MAX + 1];
+			ASSERT(tokenAmount->assetNameSize + 1 <= SIZEOF(name));
+			for (size_t i = 0; i < tokenAmount->assetNameSize; i++)
+				name[i] = tokenAmount->assetName[i];
+			name[tokenAmount->assetNameSize] = '\0';
+
+			ui_displayPaginatedText(
+			        "Token name",
```

Comment:
I'd display `Token name (ASCII)` or `Token name (text)` otherwise it opens up space for confusion should the token's name happen to be a valid hex string